zebROS_ws/src/behaviors/src/2023_fourber_server.cpp

```diff
@@ -0,0 +1,383 @@
+    // we are trying to go to an illegal position, 
+    else if (fourbar_cur_setpoint_ < min_distance) {
+      safety_req.request.position = min_distance;
+    }
+    // currentlly in illegal position but want to go to legal position
+    else if (fourbar_cur_setpoint_ >= min_distance && fourbar_cur_position_ < min_distance) {
+      safety_req.request.position = fourbar_cur_setpoint_;
+    }
+
+    safety_req.request.below = below;
+    fourbar_srv_.call(safety_req);
+    if (!waitForFourbar(safety_req.request.position)) {
+      FourberERR("Failed calling fourber with message 'safety_req'");
+      publishFailure();
+    }
+    else {
+      publishSuccess();
+    }
+  }
+
+  void publishFailure() {
+    feedback_.success = false;
+    result_.success = false;
+    as_.publishFeedback(feedback_);
+    as_.setAborted(result_);
+  }
+
+  void publishSuccess() {
+    feedback_.success = true;
+    result_.success = true;
+    as_.publishFeedback(feedback_);
+    as_.setSucceeded(result_);
+  }
+
+  void executeCB(const behavior_actions::Fourber2023GoalConstPtr &goal)
+  {
+    ros::spinOnce();
+
+    // case where elevater node has given info that we are in a zone we need to be safe
+    if (goal->safety_position == fourber_ns::SAFETY_HIGH || goal->safety_position == fourber_ns::SAFETY_INTAKE_LOW) {
+      // rememeber where we were when moving ot saftey mode as the only one that uses saftey mode is the elevatER
+      previous_setpoint_ = fourbar_cur_setpoint_;
+      if (goal->safety_position == fourber_ns::SAFETY_HIGH) {
+        FourberINFO("Safey HIGH mode called for fourbar");
+        saftey_state_ = SafteyState::SAFTEY_HIGH;
+        safetyBoundsAndCallService(safety_high_distance_, safety_high_below_);
+      }
+
+      if (goal->safety_position == fourber_ns::SAFETY_INTAKE_LOW) {
+        FourberINFO("Safey LOW mode called for fourbar");
+        saftey_state_ = SafteyState::SAFTEY_LOW;
+        safetyBoundsAndCallService(safety_low_distance_, safety_low_below_); 
+      }
+      return;
+    }
+
+    // case where we have move back to where we were
+    if (goal->safety_position == fourber_ns::SAFETY_TO_NO_SAFETY) {
+      saftey_state_ == SafteyState::NONE;
+      controllers_2023_msgs::FourBarSrv go_to_previous_req;
+      go_to_previous_req.request.position = previous_setpoint_;
+      // should be good to hardcode this because the only time we want this to be true is when we are intaking which is in a restricted zone
+      go_to_previous_req.request.below = false; 
+
+      if (!fourbar_srv_.call(go_to_previous_req)) {
+        FourberERR("Failed calling fourber with message 'go_to_previous_req'");
+        publishFailure();
+      }
+
+      if (!waitForFourbar(go_to_previous_req.request.position)) {
+        publishFailure();
+      }
+
+      publishSuccess(); 
+      return;
+    }
+
+    // select piece, nice synatax makes loading params worth it
+    double req_position = game_piece_lookup_[goal->piece][goal->mode].first;
```

Comment:
I'd error check these lookups - if they're out of bounds for some reason the server will crash.